### DISCUSSION TITLE
Primary window closes when I call a different function.

### DISCUSSION THREAD

**Discussion by jtreagan** [fltk-rs/fltk-rs#1596] [Q&A] [answered]
I'm trying to find a way to ask this question without passing on piles of code that you will have to wade through, but I have to pass on some code, so I'll pare it down to the essentials as well as I can.  I start by calling a primary window with a menu that, basically, controls the flow of the program.  Here's the code that opens the primary window:
```
fn main() {
    let app = App::default();

    let mut primarywin = Window::default().with_size(800, 500).with_pos(50,200);
    set_font_size(20);
    primarywin.set_color(Color::Blue);
    primarywin.make_resizable(true);

    qbnk_menubar(&app);

    primarywin.end();
    primarywin.show();

    app.run().unwrap();
}
```
And here's the function that adds the menu bar to that window (I deleted the menu options that aren't involved in this discussion):
```
    pub fn qbnk_menubar(app: &App) -> menu::MenuBar {
        let mut menubar = menu::MenuBar::new(0, 0, 800, 40, "");

        let app_clone = app.clone();

        menubar.add(
            "Question/New\t",
            Shortcut::None,
            menu::MenuFlag::Normal,
            move |_| qst_create(&app_clone),
        );

        menubar
    }
```
When the menu item `Question/New` is chosen, the menu calls the `qst_create()` function.  Here is the code for that function:
```
    pub fn qst_create(app: &App) {
        let mut newquest = Question::new();
        let nowtext = qst_editor(newquest.qtext.as_str(), "Question Editor");

 // Other tasks and save
    }
```
Notice that the line
```
        let nowtext = qst_editor(newquest.qtext.as_str(), "Question Editor");
```
calls a text editor.  It has its own menu that looks like this:
```
    pub fn qst_editor_menubar(edtr: &TextEditor, buf: &mut TextBuffer) -> menu::MenuBar {
        let mut menubar = menu::MenuBar::new(0, 0, 800, 40, "");

        let edtr_clone = edtr.clone();
        let mut buf_clone = buf.clone();
        menubar.add(
            "Insert_Variable\t",
            Shortcut::None,
            menu::MenuFlag::Normal,
            move |_| {
                let newtext  = // Function to create a string to be used as replacement text.
                replace_highlighted_text(&edtr_clone, &mut buf_clone, &newtext);
            },
        );
        menubar
    }
```
the final code you need to see is this function:
```
    pub fn replace_highlighted_text(edtr: &TextEditor, buf: &mut TextBuffer, rpltxt: &str) {
        let (x, y) = match edtr.buffer().unwrap().selection_position() {
            Some(position) => position,
            None => panic!("\nError!  Could not find a cursor position in the editor.\n"),
        };

        buf.remove(x, y);                         // Remove the selected text
        buf.insert(x, rpltxt);                    // Insert new text and
        edtr.buffer().unwrap().unselect();        // Unhighlight text
    }
```
This code isn't runable.  To do that I'd have to include a lot more code and I'm sure you don't want me to do that. 

Here's the problem I can't seem to solve:  When the editor is running and I highlight some text followed up by choosing the `Insert Variable` option from the editor's menu I expected the program to return to displaying the editor window with the primary window hanging around in the background.  Instead, both the editor window and the primary window close and the program moves on to the `Other Tasks` of the `qst_create()` function.  Anyone care to tell me what I'm doing wrong?  Thanks.

**Comment by MoAlyousef** [comment]
Unfortunately nothing seems suspect in the posted code. 
Are you running on windows? It could a panic, sometimes if building with an IDE, the panic and unwrap messages aren't shown.
I've tried to fill some of the gaps in the code to get something runnable:
```rust
use fltk::{prelude::*, enums::*, window::*, app::*, text::*, menu::{self, MenuBar}};

struct Question {
    qtext: String,
}

impl Question {
    pub fn new() -> Self {
        Self {
            qtext: String::new()
        }
    }
}

pub fn replace_highlighted_text(edtr: &TextEditor, buf: &mut TextBuffer, rpltxt: &str) {
    let (x, y) = match edtr.buffer().unwrap().selection_position() {
        Some(position) => position,
        None => panic!("\nError!  Could not find a cursor position in the editor.\n"),
    };

    buf.remove(x, y);                         // Remove the selected text
    buf.insert(x, rpltxt);                    // Insert new text and
    edtr.buffer().unwrap().unselect();        // Unhighlight text
}

pub fn qst_editor_menubar(edtr: &TextEditor, buf: &mut TextBuffer) -> menu::MenuBar {
    let mut menubar = menu::MenuBar::new(0, 0, 800, 40, "");

    let edtr_clone = edtr.clone();
    let mut buf_clone = buf.clone();
    menubar.add(
        "Insert_Variable\t",
        Shortcut::None,
        menu::MenuFlag::Normal,
        move |_| {
            let newtext  = "Hello World";
            replace_highlighted_text(&edtr_clone, &mut buf_clone, &newtext);
        },
    );
    menubar
}

pub fn qst_editor(s: &str, title: &str) -> TextEditor {
    let mut win = Window::default().with_size(800, 300).with_label(s);
    let mut buf = TextBuffer::default();
    let mut ed = TextEditor::new(0, 40, 800, 300 - 40, None);
    ed.set_buffer(buf.clone());
    qst_editor_menubar(&ed, &mut buf);
    win.end();
    win.show();
    ed
}

pub fn qst_create(app: &App) {
    let mut newquest = Question::new();
    let nowtext = qst_editor(newquest.qtext.as_str(), "Question Editor");
// Other tasks and save
}

pub fn qbnk_menubar(app: &App) -> menu::MenuBar {
    let mut menubar = menu::MenuBar::new(0, 0, 800, 40, "");

    let app_clone = app.clone();

    menubar.add(
        "Question/New\t",
        Shortcut::None,
        menu::MenuFlag::Normal,
        move |_| qst_create(&app_clone),
    );

    menubar
}

fn main() {
    let app = App::default();

    let mut primarywin = Window::default().with_size(800, 500).with_pos(50,200);
    set_font_size(20);
    primarywin.set_color(Color::Blue);
    primarywin.make_resizable(true);

    qbnk_menubar(&app);

    primarywin.end();
    primarywin.show();

    app.run().unwrap();
}
```
I get a panic when there is no selection and I click the Insert Variable menu item, but that's expected from the call to `panic!` in the code, otherwise it works fine. 

I also wonder why you have an app_clone, are you inadvetently calling some other app method?

**Reply by jtreagan** [reply]
@MoAlyousef ,
     I should have posted runable code.  Sorry to put you to that much work.  I'm running on Linux.  I'm using the `app_clone` because when I tried to pass `app` to `qst_create(&app)`  in the menu closure the compiler kept telling me it was escaping the closure.  The `clone` thing is my workaround.  Hmmm.....let me play around with what you have here and I'll get back to you.  I already see one change you made that I want to incorporate in my code, although that isn't going to solve the problem.

**Reply by jtreagan** [reply]
Well, I've traced it down to where the error is happening in the code, but I still don't understand why it's happening or how to correct it.  In one of the code snippets posted above, `( qst_editor_menubar() )` , I left out a function that created the replacement text.  Here's that line:
```
                let newtext  = // Function to create a string to be used as replacement text.
```
You replaced that line with 
```
            let newtext  = "Hello World";
```
Here is the original line from my project code:
```
                let newtext  = qst_make_var_replace_text();
```
The function `qst_make_var_replace_text()` calls up FLTK's file dialog, to choose a directory and then create a radiobutton menu of the file names.  It finally returns, as a String, the label of whatever button was chosen by the user.  Here's the code of that function:
```
    pub fn qst_make_var_replace_text() -> String {
        println!("\n Please choose the variable you want to insert. \n");
        let path = file_pathonly();
        let flist = file_get_dir_list(&path);
        let varname = radio_lightbtn_menu(&flist);
        let rpltxt = format!("§{}§", varname);

        rpltxt
    }
```
**Note**:  This function relies on functions from two crates I created and included in the Cargo.toml file.  I'll post the code of those functions if you need them.  Let me know.

All that function does is return a string.  **When I comment out** the call to that function and use some different string, the program returns to the editor with the replacement text displayed right where it should be.  It then waits for more input from the user.  **When I include that function**, while it _does_ replace the text with a properly formatted string, it then closes the editor and moves to the next line of code instead of waiting for more user input as it should.  And...I really don't see where my problem is.

**Reply by jtreagan** [reply]
I think you are going to need to see the code for those other files, so here it is:
```
    pub fn file_pathonly() -> String {
        let mut dialog = dialog::NativeFileChooser::new(dialog::NativeFileChooserType::BrowseFile);
        dialog.show();

        let path = dialog.filename();
        let pathonly = path.parent().unwrap().to_str().unwrap().to_string();

        pathonly
    }

    pub fn file_get_dir_list(path: &str) -> Vec<String> {
        let dir_entries = fs::read_dir(path).unwrap();

        let file_names: Vec<String> = dir_entries
            .filter_map(Result::ok)
            .filter(|entry| entry.file_type().unwrap().is_file())
            .map(|entry| entry.file_name().into_string().unwrap())
            .collect();

        file_names
    }

    pub fn radio_lightbtn_menu(flist: &Vec<String>) -> String {

        let newstring: RefCell<String> = RefCell::new("".to_string());
        let keepers: Rc<RefCell<String>> = Rc::new(newstring);

        let app = app::App::default();
        let mut win = window::Window::default().with_size(400, 300);
        let mut flex = group::Flex::default().with_size(250, 300);
        let scroll = group::Scroll::default().with_size(200, 200);
        let pack = group::Pack::default().with_size(200, 200);

        for file in flist {
            let _radio = button::RadioLightButton::default()
                .with_label(file)
                .with_size(0, 30);
        }

        pack.end();
        scroll.end();
        flex.end();

        let mut submit = button::Button::new(300, 175, 75, 30, "Submit");

        win.end();
        win.show();

        let mut keepers_clone = Rc::clone(&keepers);

        submit.set_callback(move |_b| {
            for i in 0..pack.children() {
                let radio: button::RadioLightButton = button::RadioLightButton::from_dyn_widget(&pack.child(i).unwrap()).unwrap();
                if radio.is_toggled() {
                    *keepers_clone.borrow_mut() = radio.label().clone();
                }
            }
            app.quit();
        });

        app.run().unwrap();

        let ret: String = keepers.borrow().clone();
        ret
    }
```
Hope that helps.

**Reply by MoAlyousef** [reply, marked as answer]
I can't check right now, but you shouldn't create a another app instance in radio_lightbtn_menu(), nor should you call run() or quit() on it since it basically all refer back to the same App instance. App::quit() iterates all windows in your application and closes them:
https://docs.rs/fltk/latest/src/fltk/app/rt.rs.html#146

If you need the window created there to be shown while a user selects an item, you can do:
```
        while win.shown() {
            app::wait();
        }

        let ret: String = keepers.borrow().clone();
        ret
```
And if you need the submit button to close the window, you should pass the win to the closure and hide it there.

**Reply by jtreagan** [reply]
That worked.  I've been overusing instances of `app::App`.  Removing the instances I don't need and inserting your `app::wait()` loop seems to have fixed things.  I'm likely going to have to go through my code and fix some more of those unneeded `app::App` instances.  I also think I've been thinking too linearly, not understanding how callbacks work outside/independent of the normal line by line sequence of code execution.  Am I right that `asynchronous` is the right term for that?  Anyway, @MoAlyousef , thanks for helping me out.  :>)

**Reply by MoAlyousef** [reply]
Most welcome. It is indeed asynchronous event-based.
